File: ios/ParkingPin/Identity/FirebaseAccountLinking.swift
import FirebaseAuth
import Foundation

/// The one Firebase Auth call that attaches a provider to the account this device already
/// has (docs/07 §13a).
///
/// **`link(with:)`, never `signIn(with:)`.** Linking keeps the uid and attaches a provider
/// to it; signing in mints a new uid and abandons the anonymous one together with anything
/// keyed to it. There is no server state to lose today, which is exactly why the rule is
/// enforced here rather than remembered later — the failure is invisible until it is
/// expensive. Android's counterpart was measured doing this correctly on a real device:
/// same uid before and after, one user in the project.
///
/// Thin by design, like `FirebaseAnonymousSignIn` beside it: the ordering and the
/// uid-changed check live in `LinkingAccountIdentity`, where a test can reach them.
struct FirebaseAccountLinking: AccountLinking {
    #if PK_DEV
        /// The last raw code `link(with:)` refused with, for the diagnostics file.
        ///
        /// The mapped case says which *kind* of refusal it was; this says exactly which
        /// code produced it, because on 2026-09-21 a device reported a refusal that no
        /// account in the project could explain, and a Korean sentence read off a screen is
        /// not evidence. Compiled out of STAGING and PROD.
        nonisolated(unsafe) static var lastRefusalCode: Int?
    #endif

    private let bootstrap: FirebaseBootstrap

    init(bootstrap: FirebaseBootstrap = .shared) {
        self.bootstrap = bootstrap
    }

    /// **Never touches `Auth` before Firebase is up.** `Auth.auth()` does not return nil on
    /// an unconfigured app — it raises, and the process dies. The identity here is lazy by
    /// design (docs/04_IOS §14: home renders with no network), so on an ordinary launch
    /// nothing has started Firebase yet, and this is the first thing the settings screen
    /// asks. The screen crashed on open until this guard existed.
    ///
    /// `.none` is the honest answer either way: no Firebase, no identity.
    func currentState() -> AccountState {
        guard bootstrap.isStarted, let user = Auth.auth().currentUser else { return .none }
        let provider = user.providerData
            .lazy
            .compactMap { AccountProvider(providerId: $0.providerID) }
            .first
        guard let provider else { return .anonymous(uid: user.uid) }
        return .linked(uid: user.uid, provider: provider)
    }

    func link(_ credential: AccountCredential) async -> AccountLinkResult {
        // Started by the anonymous sign-in that `LinkingAccountIdentity` does first, so this
        // is a guard rather than the ordinary path — but an unguarded `Auth.auth()` is a
        // crash, not a nil.
        guard bootstrap.isStarted, let user = Auth.auth().currentUser else {
            return .failed(reason: "no current user")
        }
        let previousUid = user.uid
        let authCredential: AuthCredential = switch credential {
        case let .apple(idToken, rawNonce):
            // `fullName` stays nil on purpose: the request never asks for it (docs/09 §1).
            OAuthProvider.appleCredential(withIDToken: idToken, rawNonce: rawNonce, fullName: nil)
        }
        do {
            let result = try await user.link(with: authCredential)
            return .linked(uid: result.user.uid, previousUid: previousUid)
        } catch let error as NSError where Self.collisionResults[error.code] != nil {
            // docs/07 §13b, and **three different situations** that were being reported as
            // one. Telling a user their account is "in use on another device" when the real
            // problem is that their Apple ID's email belongs to a Google sign-in sends them
            // looking for a phone that has nothing to do with it.
            AppLog.identity.info("link refused: \(error.code, privacy: .public)")
            #if PK_DEV
                Self.lastRefusalCode = error.code
            #endif
            return Self.collisionResults[error.code] ?? .alreadyLinkedElsewhere
        } catch {
            // The domain and code only — an auth message can carry a token fragment or an
            // email, and docs/09 §11 keeps both out of the log.
            let nsError = error as NSError
            #if PK_DEV
                Self.lastRefusalCode = nsError.code
            #endif
            AppLog.identity.info("link failed: \(nsError.domain, privacy: .public)(\(nsError.code, privacy: .public))")
            return .failed(reason: "\(nsError.domain)(\(nsError.code))")
        }
    }

    /// The collisions `link(with:)` reports, each meaning something different to the user.
    ///
    /// `accountExistsWithDifferentCredential` is the one worth naming: it is not about this
    /// credential at all but about the **email** behind it, and a project set to "one
    /// account per email address" answers it whenever the same person uses Apple here and
    /// Google elsewhere.
    private static let collisionResults: [Int: AccountLinkResult] = [
        AuthErrorCode.credentialAlreadyInUse.rawValue: .alreadyLinkedElsewhere,
        AuthErrorCode.emailAlreadyInUse.rawValue: .emailBelongsToAnotherAccount,
        AuthErrorCode.accountExistsWithDifferentCredential.rawValue: .emailBelongsToAnotherAccount,
        AuthErrorCode.providerAlreadyLinked.rawValue: .alreadyLinkedToThisAccount
    ]

    func signOut() async {
        guard bootstrap.isStarted else { return }
        try? Auth.auth().signOut()
    }
}

/// The SDK's provider ids, mapped in exactly one place.
extension AccountProvider {
    var providerId: String {
        switch self {
        case .google: "google.com"
        case .apple: "apple.com"
        }
    }

    init?(providerId: String) {
        guard let match = AccountProvider.allCases.first(where: { $0.providerId == providerId })
        else { return nil }
        self = match
    }
}

File: android/app/src/main/java/kr/parkingpin/app/identity/FirebaseAccountLinking.kt
package kr.parkingpin.app.identity

import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthUserCollisionException
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.OAuthProvider
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.tasks.await

/**
 * The one Firebase Auth call that attaches a provider to the account this device already has
 * (docs/07 §13a).
 *
 * **`linkWithCredential`, never `signInWithCredential`.** Linking keeps the uid and attaches
 * a provider to it; signing in mints a new uid and abandons the anonymous one together with
 * anything keyed to it. There is no server state to lose today, which is exactly why the
 * rule is enforced here rather than remembered later — the failure is invisible until it is
 * expensive.
 *
 * Thin by design, like [FirebaseAnonymousSignIn] beside it: the ordering, the mutex and the
 * uid-changed check live in [LinkingAccountIdentity], where a JVM test can reach them.
 */
class FirebaseAccountLinking(private val auth: FirebaseAuth) : AccountLinking {

    override fun currentState(): AccountState {
        val user = auth.currentUser ?: return AccountState.None
        val provider = user.providerData
            .firstNotNullOfOrNull { AccountProvider.of(it.providerId) }
            ?: return AccountState.Anonymous(user.uid)
        return AccountState.Linked(user.uid, provider)
    }

    override suspend fun link(provider: AccountProvider, credential: String): AccountLinkResult {
        val user = auth.currentUser ?: return AccountLinkResult.Failed("no current user")
        val previousUid = user.uid
        return try {
            val linked = user.linkWithCredential(provider.credential(credential)).await()
            val uid = linked.user?.uid ?: previousUid
            AccountLinkResult.Linked(uid = uid, previousUid = previousUid)
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (collision: FirebaseAuthUserCollisionException) {
            // docs/07 §13b, and three situations rather than one. Refused rather than
            // switched in every case — the alternative abandons this device's uid silently,
            // and v1 has nothing on the server worth that — but a user told their account
            // is "in use on another device" when the real problem is their email goes
            // looking for a phone that has nothing to do with it.
            Log.i(TAG, "link refused: ${collision.errorCode}")
            when (collision.errorCode) {
                ERROR_PROVIDER_ALREADY_LINKED -> AccountLinkResult.AlreadyLinkedToThisAccount
                ERROR_EMAIL_ALREADY_IN_USE,
                ERROR_ACCOUNT_EXISTS_WITH_DIFFERENT_CREDENTIAL,
                -> AccountLinkResult.EmailBelongsToAnotherAccount
                else -> AccountLinkResult.AlreadyLinkedElsewhere
            }
        } catch (failure: Exception) {
            // Class name only — an auth message can carry a project or a token fragment,
            // and docs/09 §11 keeps those out of the log.
            Log.i(TAG, "link failed: ${failure.javaClass.simpleName}")
            AccountLinkResult.Failed(failure.javaClass.simpleName)
        }
    }

    override suspend fun signOut() {
        auth.signOut()
    }

    private companion object {
        const val TAG = "PkAccount"

        // `FirebaseAuthUserCollisionException.errorCode`, which is a string rather than the
        // numeric code the iOS SDK reports. The two platforms answer the same three cases.
        const val ERROR_PROVIDER_ALREADY_LINKED = "ERROR_PROVIDER_ALREADY_LINKED"
        const val ERROR_EMAIL_ALREADY_IN_USE = "ERROR_EMAIL_ALREADY_IN_USE"
        const val ERROR_ACCOUNT_EXISTS_WITH_DIFFERENT_CREDENTIAL =
            "ERROR_ACCOUNT_EXISTS_WITH_DIFFERENT_CREDENTIAL"
    }
}

/**
 * The SDK's provider ids, mapped in exactly one place.
 *
 * Apple arrives as a generic OAuth credential rather than a dedicated provider class, which
 * is why the two branches do not look alike.
 */
internal fun AccountProvider.credential(token: String) = when (this) {
    AccountProvider.GOOGLE -> GoogleAuthProvider.getCredential(token, null)
    AccountProvider.APPLE -> OAuthProvider.newCredentialBuilder(APPLE_PROVIDER_ID)
        .setIdToken(token)
        .build()
}

internal fun AccountProvider.Companion.of(providerId: String): AccountProvider? = when (providerId) {
    GoogleAuthProvider.PROVIDER_ID -> AccountProvider.GOOGLE
    APPLE_PROVIDER_ID -> AccountProvider.APPLE
    else -> null
}

private const val APPLE_PROVIDER_ID = "apple.com"
